### DISCUSSION TITLE
PyInstaller generated .app file opens at terminal but not when double-clicked

### DISCUSSION THREAD

**Discussion by HenestrosaDev** [pyinstaller/pyinstaller#8773] [Help] [answered]
When I run `pyinstaller audiotext.spec` on Windows, the `.exe` file is successfully generated and I can open it without any problem. When I run this command on macOS, the bundle is also generated, but when I double-click on the `.app` file, the program doesn't start. However, if I double-click on the Unix executable file or run `open Audiotext.app`, the program starts. What is wrong with the `.app` file?

Here is the `.spec` file:

```python
# -*- mode: python ; coding: utf-8 -*-
import os.path
from pathlib import Path
from PyInstaller.compat import is_darwin, is_win

import sys ; sys.setrecursionlimit(sys.getrecursionlimit() * 5)

relative_site_packages_path = "venv/Lib/site-packages" if is_win else "venv/lib/python3.10/site-packages"

datas = [
    (f"{relative_site_packages_path}/customtkinter", "customtkinter"),
    (f"{relative_site_packages_path}/transformers", "transformers"),
    (f"{relative_site_packages_path}/lightning", "lightning"),
    (f"{relative_site_packages_path}/lightning_fabric", "lightning_fabric"),
    (f"{relative_site_packages_path}/speechbrain", "speechbrain"),
    (f"{relative_site_packages_path}/pyannote", "pyannote"),
    (f"{relative_site_packages_path}/asteroid_filterbanks", "asteroid_filterbanks"),
    (f"{relative_site_packages_path}/whisperx", "whisperx"),
    (f"{relative_site_packages_path}/librosa", "librosa"),
    ("res", "res"),
    ("config.ini", "."),
    (".env", "."),
]

hiddenimports = [
    "huggingface_hub.repository",
    "sklearn.utils._cython_blas",
    "sklearn.neighbors.quad_tree",
    "sklearn.tree",
    "sklearn.tree._utils",
]

block_cipher = None

is_debug = True
if is_debug:
    options = [("v", None, "OPTION")]
else:
    options = []

a = Analysis(
    ["src/app.py"],
    pathex=[relative_site_packages_path],
    binaries=[],
    datas=datas,
    hiddenimports=hiddenimports,
    hookspath=[],
    hooksconfig={},
    runtime_hooks=[],
    excludes=[],
    win_no_prefer_redirects=False,
    win_private_assemblies=False,
    cipher=block_cipher,
    noarchive=False,
)

pyz = PYZ(a.pure, a.zipped_data, cipher=block_cipher)

macos_icon = "res/macos/icon.icns"
windows_icon = "res/windows/icon.ico"

exe = EXE(
    pyz,
    a.scripts,
    options,
    exclude_binaries=True,
    name="Audiotext",
    debug=is_debug,
    bootloader_ignore_signals=False,
    strip=False,
    upx=True,
    upx_exclude=[],
    runtime_tmpdir=None,
    console=is_debug,
    disable_windowed_traceback=False,
    argv_emulation=False,
    target_arch=None,
    codesign_identity=None,
    entitlements_file="res/macos/entitlements.plist" if is_darwin else None,
    icon=windows_icon if is_win else macos_icon,
)

coll = COLLECT(
    exe,
    a.binaries,
    a.zipfiles,
    a.datas,
    strip=False,
    upx=True,
    upx_exclude=[],
    name="Audiotext",
)

app = BUNDLE(
    coll,
    name="Audiotext.app",
    icon=macos_icon,
    bundle_identifier="com.henestrosadev.audiotext",
    version="2.3.0",
    info_plist={
        "NSPrincipalClass": "NSApplication",
        "NSAppleScriptEnabed": False,
        "NSHighResolutionCapable": True,
        "NSMicrophoneUsageDescription": "Allow Audiotext to record audio from your microphone to generate transcriptions.",
    }
)
```

[Here](https://github.com/HenestrosaDev/audiotext) is the repo with the code (the `.spec` file and the structure of the `res` folder will not be updated until I resolve this issue).

**Comment by rokm** [comment]
> when I double-click on the .app file, the program doesn't start. However, if I double-click on the Unix executable file or run open Audiotext.app, the program starts. 

This usually indicates that you are using some functionality that implicitly depends on full shell environment, which you get when you run your app from terminal, but not when application is ran from Finder. Usually this is debugged by wrapping your *whole* entry-point script into a try/except block and write the exception to a text file at fixed location.

> ```
>     console=is_debug,
> ```

I've also seen .app bundles crash because they were not built with windowed/noconsole bootloader variant (https://github.com/pyinstaller/pyinstaller/issues/7123#issuecomment-1267549677). I.e., for .app bundles, `console` should always be set to `False`.

**Comment by bwoodsend** [comment]
I suspect you're in the same boat as https://github.com/pyinstaller/pyinstaller/issues/7459. If that's the case, you should be able to reproduce the issue by removing custom environment variables:
```bash
env -i ./dist/Audiotext.app/Contents/MacOS/Audiotext
```
If that's not the case then wrap your code in some *try/except log error* handling.
```python
import traceback

try:
    main code here (including imports)
except:
    with open("/Users/Shared/error.log", "w") as f:
        f.write(traceback.format_exc())
```

**Comment by HenestrosaDev** [comment]
I was wrapping the code without the imports inside the `try` block. Now that I've added them, I finally have the following error in the `error.log` file:

```
ERROR:root:No ffmpeg exe could be found. Install ffmpeg on your system, or set the IMAGEIO_FFMPEG_EXE environment variable.
```

ffmpeg is indeed installed on my Mac, but when I run the `.app` file, it doesn't seem to find it.

**Reply by rokm** [reply]
Like I said, .app bundles launched from Finder receive very limited shell environment. One of consequences of that is that Homebrew directories are not in `PATH`, so utilities from there cannot be found. 

You'll probably need to think what to do about `ffmpeg` anyway, if you want your .app to be portable. Either you'll need to bundle the copy from Homebrew on build system (and set `IMAGEIO_FFMPEG_EXE` so that it is discovered), or collect the copy bundled with `imageio-ffmpeg` (x86_64 macOS wheels for imageio-ffmpeg seem to bundle a ffmpeg executable), or assume that it is available on the system and modify the search paths accordingly.

**Reply by rokm** [reply]
> or collect the copy bundled with imageio-ffmpeg (x86_64 macOS wheels for imageio-ffmpeg seem to bundle a ffmpeg executable)

Actually, this should be happening [automatically](https://github.com/pyinstaller/pyinstaller-hooks-contrib/blob/master/_pyinstaller_hooks_contrib/stdhooks/hook-imageio_ffmpeg.py), so if you are running into ffmpeg-not-found with `imageio-ffmpeg`, I suppose you are using `arm64` environment?

**Reply by HenestrosaDev** [reply]
Yes, I use the `arm64` environment.

I've tried doing the following in the `.spec` file to add ffmpeg with no success:

```python
import shutil

...

binaries = [
    (shutil.which("ffmpeg"), "."),
    (shutil.which("ffprobe"), "."),
]

...

a = Analysis(
    binaries=binaries,
    ...
)
```

**Reply by rokm** [reply, marked as answer]
You need to also set `IMAGEIO_FFMPEG_EXE` environment variable at the start of your program (or add `sys._MEIPASS` to `PATH` environment variable, although `IMAGEIO_FFMPEG_EXE`  should be preferable).

**Reply by HenestrosaDev** [reply]
That worked! Just to be sure, setting a custom value for `IMAGEIO_FFMPEG_EXE` only affects the program, not the system, right?

**Reply by rokm** [reply]
> That worked! Just to be sure, setting a custom value for `IMAGEIO_FFMPEG_EXE` only affects the program, not the system, right?

Whatever environment modifications you make within your program are applicable only to that process and its child processes. (That goes for both `IMAGEIO_FFMPEG_EXE` and `PATH` environment variable modifications).

**Reply by HenestrosaDev** [reply]
Okay, thank you very much! :)

**Comment by HenestrosaDev** [comment]
Thank you @rokm and @bwoodsend for your help! `.app` now opens without problems. :)

**Comment by HenestrosaDev** [comment]
As I said earlier, the `.app` file opens without problems. However, when I try to generate a transcription with the program, I get the following error:

```
Traceback (most recent call last):
  File "handlers/whisperx_handler.py", line 49, in transcribe_file
  File "whisperx/audio.py", line 61, in load_audio
    out = subprocess.run(cmd, capture_output=True, check=True).stdout
  File "subprocess.py", line 503, in run
  File "subprocess.py", line 971, in init
  File "subprocess.py", line 1847, in _execute_child
FileNotFoundError: [Errno 2] No such file or directory: 'ffmpeg'
```

This is the code of the `load_audio` method from the `whisperx/audio.py` file:

```python
def load_audio(file: str, sr: int = SAMPLE_RATE):
    try:
        # Launches a subprocess to decode audio while down-mixing and resampling as necessary.
        # Requires the ffmpeg CLI to be installed.
        cmd = [
            "ffmpeg",
            "-nostdin",
            "-threads",
            "0",
            "-i",
            file,
            "-f",
            "s16le",
            "-ac",
            "1",
            "-acodec",
            "pcm_s16le",
            "-ar",
            str(sr),
            "-",
        ]
        out = subprocess.run(cmd, capture_output=True, check=True).stdout
    except subprocess.CalledProcessError as e:
        raise RuntimeError(f"Failed to load audio: {e.stderr.decode()}") from e

    return np.frombuffer(out, np.int16).flatten().astype(np.float32) / 32768.0
```

**Reply by rokm** [reply]
This sort of invocation requires `ffmpeg` executable to be in `PATH`.

**Reply by HenestrosaDev** [reply]
Indeed. I fixed it by adding the absolute path to the root of the program. Here is the code:

```python
from pathlib import Path
import os

def get_root_path() -> Path:
    if getattr(sys, "frozen", False) and hasattr(sys, "_MEIPASS"):
        return Path(sys._MEIPASS)
    else:
        return Path(__file__).parent.parent.parent

os.environ["PATH"] += os.pathsep + str(get_root_path())
```

Once again, thank you very much!
